# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "((keyword lambda) (one or more inputs) (colon) (function to be executed))(value t pass in to the lambda)"</p>

#### Syntax

In [4]:
def addTwo(x):
    return x + 2
print(addTwo(4))

# Lambda func
a_var = 6
print((lambda x: x + 2) (a_var))
    

6
8


#### Saving to a Variable

In [6]:
f_test = lambda x: x + 2
f_test(5)

7

#### Multiple Inputs

In [8]:
# multiple inputs with no variable
print((lambda x,y,z: x * y * z)(3,5,8))


# or
x_test = lambda x,y,z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [15]:
# functional programing
def multiply(f, num):
    '''
    f expects a lambda function num expects a number
    '''
    return f(num)
multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [18]:
# getting a number back
# normal defined funct
def multiply_test(num):
    return num * 4

# func within func
def returnFunc():
    test = 4
    def multiply(num): #closure
        return num * 2
    return multiply
f_return = returnFunc()
print(returnFunc())
print(f_return(4))

# lambda function return from a regular function
def returnLam(b,c):
    return lambda x,a: x + a + b + c
r_lamb = returnLam(4,5)
print(r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x0000023D1AF587C0>
8
19


#### If Statements within Lambdas

In [11]:
# lambda x: True if (condition) else False
f_condition = lambda num : num * 2 if num > 10 else num + 2

print(f_condition(8))
print(f_condition(15))

10
30


In [24]:
# cubes arguments passed in and assign the anonymous func to a variable f
def cube(x):
    return x**3
f = lambda x : cube(x)
f(3)


27

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [31]:
# map(function, iterable)
# an iterable is something you can loop over; such as a list, dict, tuple, set or string
# normally, the usage the map happens with a pre-defined function, but we can use lambdas as well
def squared(num, num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 **2
    else:
        return num, num2
numbers = [4, 11, 20, 3, 15, 20]
more_nums = [4, 10, 3, 2, 6]
squared_nums_map = list(map(squared, numbers, more_nums))
print(squared_nums_map)

#   we are passing the func sqaured in, it will go over the numbers and more nums and run the squared func on more_nums and numbe

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [32]:
# map(lambda x: do something, iterable)
# using lambda  in map usually all takes place on one line
squared_nums_lamb = list(map(lambda x, y: (x**2, y**2) if x < 10 and y < 10 else (x,y), numbers, more_nums))
print(squared_nums_lamb)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [37]:
#  double each number 2*? -1 ?
numbers = [2, 5, 7 , 9, 11]
double_nums = list(map(lambda x: 2*x -1, numbers))
print(double_nums)


[3, 9, 13, 17, 21]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [38]:
names = ['Bob', 'Max',  'Alex']
def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
new_names_list = list(filter(a_names, names))
print(new_names_list)
# filter funcs return True/false values

['Alex']


#### Using Lambda's with Filter()

In [39]:
new_names_lamb = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lamb)

['Alex']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [ ]:
from statistics import mean
# mean is sum of all values in an entity divided by the num of val in an entity
nums = [2, 7, 4, 2, 1, 6,9, 4, 4, 9]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [ ]:
# just use a for loop

#### Using Lambda's with Reduce()

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [40]:
# def addTwo():
#     return addTwo(2)
# not best pratice as it will never break and keep returning over and over 

# do this

def addNums(num):
#     set a base case to stop recurssion from going forever!!
    if num <= 1: #tricky..knowing whether or not you're going to add 0 or 1
        print("addNums(1) = 1")
        return num
    else:
        print(f'addNums({num} = {num}) + addNums({num - 1})')
        return num + addNums(num - 1)
addNums(5)

addNums(5 = 5 + addNums(4))
addNums(4 = 4 + addNums(3))
addNums(3 = 3 + addNums(2))
addNums(2 = 2 + addNums(1))
addNums(1) = 1


15

#### Writing a Factorial Function

In [41]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)
factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [47]:
# desired output = 
# SubNums(5) = 5 - subNums(4)
# subNums(4 ) = 4 -subNums(3)
# ...
# ..
# subNum(1) = 1
# pseudocode: Function subNums takes in a number 'num'
# If 'num' is less than or equal to 1
#  Print 'subNums(1) = 1'
#  Return 'num'
#  Else
#  Print 'subNums('num') = 'num' - subNums('num' - 1)'
#  Return 'num' - Call subNums with ('num' - 1)
# End Function

# Call subNums with 5

def subNums(num):
    if num <= 1:
        print('subNums(1) = 1')
        return num
    else:
         print(f'subNums({num} = {num}) - subNums({num - 1})')
         return num - subNums(num - 1)
subNums(5)

subNums(5 = 5) - subNums(4)
subNums(4 = 4) - subNums(3)
subNums(3 = 3) - subNums(2)
subNums(2 = 2) - subNums(1)
subNums(1) = 1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [48]:
# making our own range of func
def my_range(stop,start = 0,step = 2):
    while start < stop:
        yield start #if you use return here, it will only print the first num
        start += step
for i in my_range(20):
    my_generator_value = i
    print(my_generator_value)

0
2
4
6
8
10
12
14
16
18


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [69]:
# takes num and yields num **2 , prints each num squared until zero?
# starts with the num (start)
# squares it and yields the results
# decreases start by step until start is 0/less than stop
# #pseudocode: 
# squared_num(start, step, stop = 0):
#     While start is greater than or equal to stop:
#         Yield the square of start
#         Decrement start by step

# End 

# using generator: gen_value
#     For i in squared_num(10, 2):
#         Print gen_value = i

def squared_num(start, step, stop = 0):
    while start >= stop:
        yield start ** 2
        start -= step
        
#  using generator
for i in squared_num(10, 2):
    my_gen_value = i
    print(my_gen_value)

100
64
36
16
4
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [53]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

# filter out elements that are not empty and do not start with a space
# remove trailing spaces with .strip()
# lambda check if place is not an empty string after removing trailing spaces
# then filter to iterate over each element and return a new list of places.
# SOLUTION:

filtered_strings = list(filter(lambda place: place.strip() != "", places))
print(filtered_strings)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [54]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
# use what ? anonymous func = lambda func
# for what? to sort list by the last name 
# how? Pseudocode:
# step 1: define a function sort_by_last_name(author)
# sort author.. use .sort() then .split()?
# key argument = lambda name: name.split() [-1].lower()
# [-1] takes the last word in string (assuming string is first and last names).. presenting a reversed string of names.. should I use .join?

def sort_by_last_name(author):
    author.sort(key = lambda name: name.split() [-1].lower())
    return author

sorted_author = sort_by_last_name(author)
print(sorted_author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [59]:
# F = (9/5)*C + 32
# using map with lambda
# using lambda with map takes place in one line
# list contains tuples and place, then Celcius value .. use index
# pseudocode: function celcius_farhenheit =
# take a the list (places),(a,b) for each tuple in place, extract the name(a[0]), keep the same
# now extract Celcius (b[1]) amd convert to farhenheit [a[0], b[1] * (9/5)*C + 32]
# print
# SOLUTION:
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

celcius_farhenheit = list(map(lambda place: (place[0], place[1] * 9/5 + 32), places))
print(celcius_farhenheit)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [61]:
# fibonacci sequence? = a series of nums in which each num is the sum of the two preceding ones..
# often starts with 0 and 1, or 1 and 1
# what? write a recursion function to perform fib sequence
# for the desired output, seq starts with 1, 1
# base case: if num 


In [62]:
def fibonacci_seq(n):
#     base case
    if n <= 1:
        return 1
#    recursive case if n is not 1 then..
    else:
        return fibonacci_seq(n-1) + fibonacci_seq(n-2) # add preceding 2 values 


def fib(n):
    for i in range(n+1):
        print(f'Iteration {i}: {fibonacci_seq(i)}')
        
fib(5)

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
